File: U2/TG2.2 PAO202551_APIGOOGLEMAPS_EUREKABANK_RESFUL_DONET_GR01/01 SERVIDOR/EurekaBank_RestFull_DotNet_GR01/EurekaBank_RestFull_DotNet_GR01/DAL/SucursalDAO.cs
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using System.Linq;
using Dapper;
using EurekaBank_RestFull_DotNet_GR01.Models;

namespace EurekaBank_RestFull_DotNet_GR01.DAL
{
    /// <summary>
    /// Data Access Object para la entidad Sucursal usando Dapper
    /// </summary>
    public class SucursalDAO
    {
        /// <summary>
        /// Obtiene una sucursal por su código
        /// </summary>
        /// <param name="codigo">Código de la sucursal</param>
        /// <returns>Sucursal encontrada o null</returns>
        public Sucursal ObtenerPorCodigo(int codigo)
        {
            try
            {
                using (var conn = ConexionDB.ObtenerConexion())
                {
                    string query = @"SELECT 
                                    chr_sucucodigo AS Codigo,
                                    vch_sucunombre AS Nombre,
                                    vch_sucuciudad AS Ciudad,
                                    vch_sucudireccion AS Direccion,
                                    int_sucucontcuenta AS ContadorCuentas,
                                    dec_suculatitud AS Latitud,
                                    dec_suculongitud AS Longitud
                                    FROM Sucursal 
                                    WHERE chr_sucucodigo = @Codigo";
                    
                    return conn.QueryFirstOrDefault<Sucursal>(query, new { Codigo = codigo });
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al obtener sucursal: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Lista todas las sucursales
        /// </summary>
        /// <returns>Lista de todas las sucursales</returns>
        public List<Sucursal> ListarTodas()
        {
            try
            {
                using (var conn = ConexionDB.ObtenerConexion())
                {
                    string query = @"SELECT 
                                    chr_sucucodigo AS Codigo,
                                    vch_sucunombre AS Nombre,
                                    vch_sucuciudad AS Ciudad,
                                    vch_sucudireccion AS Direccion,
                                    int_sucucontcuenta AS ContadorCuentas,
                                    dec_suculatitud AS Latitud,
                                    dec_suculongitud AS Longitud
                                    FROM Sucursal 
                                    ORDER BY vch_sucunombre";
                    
                    return conn.Query<Sucursal>(query).ToList();
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al listar sucursales: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Lista todas las sucursales que tienen coordenadas de geolocalización
        /// </summary>
        /// <returns>Lista de sucursales con coordenadas</returns>
        public List<Sucursal> ListarConCoordenadas()
        {
            try
            {
                using (var conn = ConexionDB.ObtenerConexion())
                {
                    string query = @"SELECT 
                                    chr_sucucodigo AS Codigo,
                                    vch_sucunombre AS Nombre,
                                    vch_sucuciudad AS Ciudad,
                                    vch_sucudireccion AS Direccion,
                                    int_sucucontcuenta AS ContadorCuentas,
                                    dec_suculatitud AS Latitud,
                                    dec_suculongitud AS Longitud
                                    FROM Sucursal 
                                    WHERE dec_suculatitud IS NOT NULL 
                                    AND dec_suculongitud IS NOT NULL
                                    ORDER BY vch_sucunombre";
                    
                    return conn.Query<Sucursal>(query).ToList();
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al listar sucursales con coordenadas: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Lista las sucursales por ciudad
        /// </summary>
        /// <param name="ciudad">Nombre de la ciudad</param>
        /// <returns>Lista de sucursales en la ciudad especificada</returns>
        public List<Sucursal> ListarPorCiudad(string ciudad)
        {
            try
            {
                using (var conn = ConexionDB.ObtenerConexion())
                {
                    string query = @"SELECT 
                                    chr_sucucodigo AS Codigo,
                                    vch_sucunombre AS Nombre,
                                    vch_sucuciudad AS Ciudad,
                                    vch_sucudireccion AS Direccion,
                                    int_sucucontcuenta AS ContadorCuentas,
                                    dec_suculatitud AS Latitud,
                                    dec_suculongitud AS Longitud
                                    FROM Sucursal 
                                    WHERE UPPER(vch_sucuciudad) = UPPER(@Ciudad)
                                    ORDER BY vch_sucunombre";
                    
                    return conn.Query<Sucursal>(query, new { Ciudad = ciudad }).ToList();
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al listar sucursales por ciudad: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Actualiza las coordenadas de una sucursal
        /// </summary>
        /// <param name="codigo">Código de la sucursal</param>
        /// <param name="latitud">Nueva latitud</param>
        /// <param name="longitud">Nueva longitud</param>
        /// <returns>True si se actualizó correctamente</returns>
        public bool ActualizarCoordenadas(int codigo, decimal? latitud, decimal? longitud)
        {
            try
            {
                using (var conn = ConexionDB.ObtenerConexion())
                {
                    string query = @"UPDATE Sucursal 
                                    SET dec_suculatitud = @Latitud,
                                        dec_suculongitud = @Longitud
                                    WHERE chr_sucucodigo = @Codigo";
                    
                    int filasAfectadas = conn.Execute(query, new { 
                        Codigo = codigo, 
                        Latitud = latitud, 
                        Longitud = longitud 
                    });
                    return filasAfectadas > 0;
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al actualizar coordenadas: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Inserta una nueva sucursal
        /// </summary>
        /// <param name="sucursal">Datos de la sucursal a insertar</param>
        /// <returns>Código de la sucursal creada o 0 si falló</returns>
        public int Insertar(Sucursal sucursal)
        {
            try
            {
                using (var conn = ConexionDB.ObtenerConexion())
                {
                    string query = @"INSERT INTO Sucursal 
                                    (vch_sucunombre, vch_sucuciudad, 
                                    vch_sucudireccion, int_sucucontcuenta, dec_suculatitud, dec_suculongitud)
                                    VALUES 
                                    (@Nombre, @Ciudad, @Direccion, 
                                    @ContadorCuentas, @Latitud, @Longitud);
                                    SELECT CAST(SCOPE_IDENTITY() as int)";
                    
                    var nuevoCodigo = conn.QuerySingle<int>(query, sucursal);
                    return nuevoCodigo;
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al insertar sucursal: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Actualiza los datos de una sucursal
        /// </summary>
        /// <param name="sucursal">Sucursal con los datos actualizados</param>
        /// <returns>True si se actualizó correctamente</returns>
        public bool Actualizar(Sucursal sucursal)
        {
            try
            {
                using (var conn = ConexionDB.ObtenerConexion())
                {
                    string query = @"UPDATE Sucursal SET 
                                    vch_sucunombre = @Nombre,
                                    vch_sucuciudad = @Ciudad,
                                    vch_sucudireccion = @Direccion,
                                    int_sucucontcuenta = @ContadorCuentas,
                                    dec_suculatitud = @Latitud,
                                    dec_suculongitud = @Longitud
                                    WHERE chr_sucucodigo = @Codigo";
                    
                    int filasAfectadas = conn.Execute(query, sucursal);
                    return filasAfectadas > 0;
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al actualizar sucursal: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Elimina una sucursal por su código
        /// </summary>
        /// <param name="codigo">Código de la sucursal a eliminar</param>
        /// <returns>True si se eliminó correctamente</returns>
        public bool Eliminar(int codigo)
        {
            try
            {
                using (var conn = ConexionDB.ObtenerConexion())
                {
                    string query = "DELETE FROM Sucursal WHERE chr_sucucodigo = @Codigo";
                    int filasAfectadas = conn.Execute(query, new { Codigo = codigo });
                    return filasAfectadas > 0;
                }
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al eliminar sucursal: {ex.Message}", ex);
            }
        }
    }
}

File: U2/TG2.1 PAO202551_API_GOOGLEMAPS_EUREKABANK_RESTFUL_JAVA_GR01/01 SERVIDOR/EurekaBank_RestFull_Java_GR01/src/java/ec/edu/monster/dal/SucursalDAO.java
package ec.edu.monster.dal;

import ec.edu.monster.db.ConexionDB;
import ec.edu.monster.models.Sucursal;
import java.math.BigDecimal;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

/**
 * Data Access Object para la entidad Sucursal
 * 
 * @author EurekaBank
 */
public class SucursalDAO {

    /**
     * Lista todas las sucursales
     * 
     * @return Lista de sucursales
     * @throws SQLException si hay error en la BD
     */
    public List<Sucursal> listarTodas() throws SQLException {
        List<Sucursal> sucursales = new ArrayList<>();
        String query = "SELECT chr_sucucodigo, vch_sucunombre, vch_sucuciudad, "
                + "vch_sucudireccion, int_sucucontcuenta, dec_suculatitud, dec_suculongitud "
                + "FROM Sucursal";

        try (Connection conn = ConexionDB.obtenerConexion();
                PreparedStatement ps = conn.prepareStatement(query);
                ResultSet rs = ps.executeQuery()) {

            while (rs.next()) {
                Sucursal sucursal = new Sucursal();
                sucursal.setCodigo(rs.getInt("chr_sucucodigo"));
                sucursal.setNombre(rs.getString("vch_sucunombre"));
                sucursal.setCiudad(rs.getString("vch_sucuciudad"));
                sucursal.setDireccion(rs.getString("vch_sucudireccion"));
                sucursal.setContadorCuentas(rs.getInt("int_sucucontcuenta"));
                sucursal.setLatitud(rs.getBigDecimal("dec_suculatitud"));
                sucursal.setLongitud(rs.getBigDecimal("dec_suculongitud"));
                sucursales.add(sucursal);
            }
        }
        return sucursales;
    }

    /**
     * Obtiene una sucursal por su código
     * 
     * @param codigo Código de la sucursal
     * @return Sucursal encontrada o null
     * @throws SQLException si hay error en la BD
     */
    public Sucursal obtenerPorCodigo(int codigo) throws SQLException {
        String query = "SELECT chr_sucucodigo, vch_sucunombre, vch_sucuciudad, "
                + "vch_sucudireccion, int_sucucontcuenta, dec_suculatitud, dec_suculongitud "
                + "FROM Sucursal WHERE chr_sucucodigo = ?";

        try (Connection conn = ConexionDB.obtenerConexion();
                PreparedStatement ps = conn.prepareStatement(query)) {

            ps.setInt(1, codigo);
            ResultSet rs = ps.executeQuery();

            if (rs.next()) {
                Sucursal sucursal = new Sucursal();
                sucursal.setCodigo(rs.getInt("chr_sucucodigo"));
                sucursal.setNombre(rs.getString("vch_sucunombre"));
                sucursal.setCiudad(rs.getString("vch_sucuciudad"));
                sucursal.setDireccion(rs.getString("vch_sucudireccion"));
                sucursal.setContadorCuentas(rs.getInt("int_sucucontcuenta"));
                sucursal.setLatitud(rs.getBigDecimal("dec_suculatitud"));
                sucursal.setLongitud(rs.getBigDecimal("dec_suculongitud"));
                return sucursal;
            }
        }
        return null;
    }

    /**
     * Inserta una nueva sucursal
     * 
     * @param sucursal Datos de la sucursal
     * @return Código de la nueva sucursal o -1 si falla
     * @throws SQLException si hay error en la BD
     */
    public int insertar(Sucursal sucursal) throws SQLException {
        // Obtenemos el siguiente código disponible (max + 1) ya que parece no ser
        // autoincremental en la definición original
        // o usamos el contador si existe tabla de contadores para sucursal.
        // Asumiendo estrategia similar a cliente/empleado o autoincremental si se
        // ajustó BD.
        // Dado que el usuario pidió CRUD completo, asumo que debemos manejar la
        // generación de ID.
        // Revisando EmpleadoDAO se usa una tabla 'contador'. Verificaremos si hay
        // contador para Sucursal.
        // Si no, usaremos MAX+1 como fallback seguro para este ejercicio.

        int nuevoCodigo = generarNuevoCodigo();

        String query = "INSERT INTO Sucursal (chr_sucucodigo, vch_sucunombre, vch_sucuciudad, "
                + "vch_sucudireccion, int_sucucontcuenta, dec_suculatitud, dec_suculongitud) "
                + "VALUES (?, ?, ?, ?, ?, ?, ?)";

        try (Connection conn = ConexionDB.obtenerConexion();
                PreparedStatement ps = conn.prepareStatement(query)) {

            ps.setInt(1, nuevoCodigo);
            ps.setString(2, sucursal.getNombre());
            ps.setString(3, sucursal.getCiudad());
            ps.setString(4, sucursal.getDireccion());
            ps.setInt(5, 0); // Contador cuentas inicia en 0
            ps.setBigDecimal(6, sucursal.getLatitud());
            ps.setBigDecimal(7, sucursal.getLongitud());

            int filas = ps.executeUpdate();
            return filas > 0 ? nuevoCodigo : -1;
        }
    }

    private int generarNuevoCodigo() throws SQLException {
        String query = "SELECT MAX(CAST(chr_sucucodigo AS signed)) FROM sucursal"; // Cast por si acaso la DB es vieja,
                                                                                   // pero ahora es INT
        // Como la DB define chr_sucucodigo como INT, el MAX directo funciona.
        query = "SELECT MAX(chr_sucucodigo) FROM Sucursal";

        try (Connection conn = ConexionDB.obtenerConexion();
                Statement stmt = conn.createStatement();
                ResultSet rs = stmt.executeQuery(query)) {

            if (rs.next()) {
                return rs.getInt(1) + 1;
            }
        }
        return 1;
    }

    /**
     * Actualiza una sucursal existente
     * 
     * @param sucursal Datos actualizados
     * @return true si se actualizó
     * @throws SQLException si hay error en la BD
     */
    public boolean actualizar(Sucursal sucursal) throws SQLException {
        String query = "UPDATE Sucursal SET vch_sucunombre = ?, vch_sucuciudad = ?, "
                + "vch_sucudireccion = ?, dec_suculatitud = ?, dec_suculongitud = ? "
                + "WHERE chr_sucucodigo = ?";

        try (Connection conn = ConexionDB.obtenerConexion();
                PreparedStatement ps = conn.prepareStatement(query)) {

            ps.setString(1, sucursal.getNombre());
            ps.setString(2, sucursal.getCiudad());
            ps.setString(3, sucursal.getDireccion());
            ps.setBigDecimal(4, sucursal.getLatitud());
            ps.setBigDecimal(5, sucursal.getLongitud());
            ps.setInt(6, sucursal.getCodigo());

            return ps.executeUpdate() > 0;
        }
    }

    /**
     * Elimina una sucursal (marcado lógico o físico dependiendo de reglas, pero el
     * código C# dice Eliminar)
     * Generalmente en EurekaBank se usa estado, haremos borrado lógico 'ANULADO' o
     * borrado físico si no hay estado 'ELIMINADO'.
     * El query de listar filtra por 'ACTIVO', así que cambiaremos estado a
     * 'ANULADO'.
     * 
     * @param codigo Código de la sucursal
     * @return true si se eliminó
     * @throws SQLException si hay error en la BD
     */
    public boolean eliminar(int codigo) throws SQLException {
        String query = "DELETE FROM Sucursal WHERE chr_sucucodigo = ?";

        try (Connection conn = ConexionDB.obtenerConexion();
                PreparedStatement ps = conn.prepareStatement(query)) {

            ps.setInt(1, codigo);
            return ps.executeUpdate() > 0;
        }
    }

    /**
     * Actualiza solo las coordenadas
     * 
     * @param codigo   Código de la sucursal
     * @param latitud  Nueva latitud
     * @param longitud Nueva longitud
     * @return true si se actualizó
     * @throws SQLException si hay error en la BD
     */
    public boolean actualizarCoordenadas(int codigo, BigDecimal latitud, BigDecimal longitud) throws SQLException {
        String query = "UPDATE Sucursal SET dec_suculatitud = ?, dec_suculongitud = ? "
                + "WHERE chr_sucucodigo = ?";

        try (Connection conn = ConexionDB.obtenerConexion();
                PreparedStatement ps = conn.prepareStatement(query)) {

            ps.setBigDecimal(1, latitud);
            ps.setBigDecimal(2, longitud);
            ps.setInt(3, codigo);

            return ps.executeUpdate() > 0;
        }
    }
}
